cpp/src/randomforest/randomforest.cu

```diff
@@ -340,14 +393,211 @@ RF_metrics rfClassifier<T>::cross_validate(const cumlHandle& user_handle, const
 	return stats;
 }
 
-template class rf<float>;
-template class rf<double>;
+
+/**
+ * @brief Construct rfRegressor object.
+ * @tparam T: data type for input data (float or double).
+ * @param[in] cfg_rf_params: Random forest hyper-parameter struct.
+ */
+template <typename T>
+rfRegressor<T>::rfRegressor(RF_params cfg_rf_params): rf<T, T>::rf(cfg_rf_params, RF_type::REGRESSION) {};
+
+/**
+ * @brief Destructor for random forest regressor object.
+ * @tparam T: data type for input data (float or double).
+ */
+template<typename T>
+rfRegressor<T>::~rfRegressor() {
+	delete [] trees;
+}
+
+/**
+ * @brief Return a const pointer to decision trees.
+ * @tparam T: data type for input data (float or double).
+ */
+template<typename T>
+const DecisionTree::DecisionTreeRegressor<T> * rfRegressor<T>::get_trees_ptr() const {
+	return trees;
+}
+
+/**
+ * @brief Build (i.e., fit, train) random forest regressor for input data.
+ * @tparam T: data type for input data (float or double).
+ * @param[in] user_handle: cumlHandle
+ * @param[in] input: train data (n_rows samples, n_cols features) in column major format, excluding labels. Device pointer.
+ * @param[in] n_rows: number of training data samples.
+ * @param[in] n_cols: number of features (i.e., columns) excluding target feature.
+ * @param[in] labels: 1D array of target features (float or double), with one label per training sample. Device pointer.
+ */
+template <typename T>
+void rfRegressor<T>::fit(const cumlHandle& user_handle, T * input, int n_rows, int n_cols, T * labels) {
+
+	ASSERT(!trees, "Cannot fit an existing forest.");
+	ASSERT((n_rows > 0), "Invalid n_rows %d", n_rows);
+	ASSERT((n_cols > 0), "Invalid n_cols %d", n_cols);
+
+	trees = new DecisionTree::DecisionTreeRegressor<T>[this->rf_params.n_trees];
+	int n_sampled_rows = this->rf_params.rows_sample * n_rows;
+
+	const cumlHandle_impl& handle = user_handle.getImpl();
+	cudaStream_t stream = user_handle.getStream();
+
+	for (int i = 0; i < this->rf_params.n_trees; i++) {
```

Comment:
Question: If the stream from the cumlHandle was synchronized before this for loop, could `n_trees` streams be created temporarily and an `omp parallel for` used to paralellize the tree creation even further?

Reply:
ideally we could do that. But to keep in mind it comes at cost of extra memory buffers for tree building. 

As per the Developer Guide https://github.com/rapidsai/cuml/blob/branch-0.8/cpp/DEVELOPER_GUIDE.md  we need to ensure thread safety. hence a thread per loop iteration might not be the right way.

Other option is to have a loop at every kernel launch, for N streams. But that means we need to change the whole build process. This also has the same memory constraints.